File: scripts/ci/lib/llama-server-arg-rejection.mjs
#!/usr/bin/env node
/**
 * The one JS-side reader of the declared llama-server launch-argument rejection marker
 * (governance/llama-server-arg-rejection.v1.json, tempdoc 835). No copy of the string lives here:
 * the register is the source, and the runtime detector (LlamaServerOps) is pinned to the same
 * register by a Java test. Match the prefix — the suffix is per-build wording
 * ("invalid value" on b8185, "invalid stoi argument" on b8571).
 */

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const HERE = path.dirname(fileURLToPath(import.meta.url));
const REGISTER_PATH = path.resolve(HERE, '..', '..', '..', 'governance', 'llama-server-arg-rejection.v1.json');

let cachedMarker = null;

/** The declared rejection marker (prefix) for `--reasoning-budget`. Throws if the register is unreadable. */
export function reasoningBudgetRejectionMarker() {
  if (cachedMarker == null) {
    const doc = JSON.parse(fs.readFileSync(REGISTER_PATH, 'utf8'));
    const marker = doc?.reasoningBudget?.rejectionMarker;
    if (typeof marker !== 'string' || marker.length === 0) {
      throw new Error(`Missing reasoningBudget.rejectionMarker in ${REGISTER_PATH}`);
    }
    cachedMarker = marker;
  }
  return cachedMarker;
}

/** True when llama-server output shows this build rejecting the `--reasoning-budget` argument. */
export function isReasoningBudgetRejection(serverOutput) {
  if (typeof serverOutput !== 'string' || serverOutput.length === 0) return false;
  return serverOutput.includes(reasoningBudgetRejectionMarker());
}

export const REASONING_BUDGET_REGISTER_PATH = REGISTER_PATH;

File: modules/app-inference/src/main/java/io/justsearch/app/inference/LlamaServerArgRejection.java
/* SPDX-License-Identifier: Apache-2.0 */
package io.justsearch.app.inference;

/**
 * Detects llama-server rejecting a launch argument — the authoritative capability signal for flags
 * whose support varies by build (tempdoc 835 §5.2 signal 1).
 *
 * <p>Match the argument-name <em>prefix</em>, never the parser's suffix: b8185 emitted {@code
 * invalid value} and b8571 emits {@code invalid stoi argument} for the same rejection, and a
 * suffix-matching detector in a CI script silently stopped firing when the bundled build moved. The
 * marker is declared once in {@code governance/llama-server-arg-rejection.v1.json}; this constant is
 * pinned to that register by {@code LlamaServerArgRejectionContractTest}, and the CI script reads
 * the register directly, so no third copy exists.
 */
final class LlamaServerArgRejection {

  /** Declared rejection marker for {@code --reasoning-budget}. */
  static final String REASONING_BUDGET_MARKER = "error while handling argument \"--reasoning-budget\"";

  private LlamaServerArgRejection() {}

  /** True when llama-server output shows this build refusing the {@code --reasoning-budget} flag. */
  static boolean isReasoningBudgetRejection(String serverOutput) {
    return serverOutput != null && serverOutput.contains(REASONING_BUDGET_MARKER);
  }
}
